worker/upload.go

```diff
@@ -0,0 +1,1126 @@
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	// create slab id
+	var sID slabID
+	frand.Read(sID[:])
+	u.ongoing = append(u.ongoing, sID)
+
+	// create slab upload
+	slab := &slabUpload{
+		mgr: u.mgr,
+
+		upload:  u,
+		sID:     sID,
+		started: time.Now(),
+
+		overdriving: make(map[int]int, len(shards)),
+		remaining:   make(map[int]shardCtx, len(shards)),
+		sectors:     make([]object.Sector, len(shards)),
+	}
+
+	// prepare shard uploads
+	responseChan := make(chan shardResp)
+	uploads := make([]*shardUpload, len(shards))
+	for sI, shard := range shards {
+		// create the shard upload's cancel func
+		sCtx, cancel := context.WithCancel(ctx)
+		slab.remaining[sI] = shardCtx{ctx: sCtx, cancel: cancel}
+
+		// create the upload's span
+		sCtx, span := tracing.Tracer.Start(sCtx, "uploadShard")
+		span.SetAttributes(attribute.Bool("overdrive", false))
+		span.SetAttributes(attribute.Int("sector", sI))
+
+		// create the shard upload
+		uploads[sI] = &shardUpload{
+			upload:       u,
+			sID:          sID,
+			ctx:          sCtx,
+			responseChan: responseChan,
+
+			sector:      (*[rhpv2.SectorSize]byte)(shard),
+			sectorIndex: sI,
+		}
+	}
+
+	return slab, uploads, responseChan
+}
+
+func (u *upload) canUseUploader(ul *uploader, sID slabID) bool {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	_, excluded := u.excluded[ul.fcid]
+	if excluded {
+		return false
+	}
+
+	_, used := u.used[sID][ul.fcid]
+	return !used
+}
+
+func (u *upload) start(ctx context.Context, r io.Reader, rs api.RedundancySettings) chan slabResponse {
+	// create the response channel
+	slabsChan := make(chan slabResponse)
+
+	// launch a goroutine uploading the slabs
+	go func() {
+		var wg sync.WaitGroup
+		defer func() {
+			wg.Wait()
+			close(slabsChan)
+		}()
+
+		var slabIndex int
+		for res := range u.read(ctx, r, rs) {
+			if res.err != nil {
+				slabsChan <- slabResponse{err: res.err}
+				return
+			}
+
+			wg.Add(1)
+			go func(data []byte, length, index int) {
+				defer wg.Done()
+
+				// add tracing
+				ctx, span := tracing.Tracer.Start(ctx, "createShards")
+				defer span.End()
+
+				// create slab
+				s := object.NewSlab(uint8(rs.MinShards))
+
+				// create the shards
+				shards := make([][]byte, rs.TotalShards)
+				s.Encode(data, shards)
+				s.Encrypt(shards)
+
+				// upload the shards
+				var err error
+				s.Shards, err = u.uploadShards(ctx, shards, index)
+				if err != nil {
+					select {
+					case slabsChan <- slabResponse{err: err}:
+					default:
+					}
+					return
+				}
+
+				// send the slab
+				select {
+				case slabsChan <- slabResponse{
+					slab: object.SlabSlice{
+						Slab:   s,
+						Offset: 0,
+						Length: uint32(length),
+					},
+					index: index,
+				}:
+				default:
+				}
+			}(res.data, res.length, slabIndex)
+			slabIndex++
+		}
+	}()
+	return slabsChan
+}
+
+func (u *upload) registerUsedQueue(sID slabID, fcid types.FileContractID) {
```

Comment:
`Queue` is a bit misleading here. Maybe `registerUsedSlab`?

Reply:
I'll do `registerUsedUploader`. I considered `registerUsedContract` but I think uploader is probably want we want here. Just something I forgot to rename.